### PR TITLE
Secondary index order

### PR DESCRIPTION
**Pull request by ammuv**
Closed issue #220.

Added Secondary Index for Order to get Order Id using Customer Id.

Modified OrderStatus Transaction to use the secondary index.

### CODE REVIEW

**Review by jfunston — CHANGES_REQUESTED** [review]

File: src/k2/cmd/tpcc/datagen.h [hunk 1]
```diff
@@ -105,6 +105,13 @@ struct TPCCDataGen {
             data.push_back([_order=std::move(order)] (k2::K2TxnHandle& txn) mutable {
                 return writeRow<Order>(_order, txn);
             });
+
+            // populate secondary index idx_order_customer
+            auto idx_order_customer = IdxOrderCustomer(order.WarehouseID.value(), order.DistrictID.value(),
```
[the remark this answers is not in the document]
Comment (ammuv) [reply src/k2/cmd/tpcc/datagen.h:110]: done

File: src/k2/cmd/tpcc/transactions.h [hunk 2]
```diff
@@ -678,55 +684,85 @@ class OrderStatusT : public TPCCTxn
...
                 return make_ready_future();
             });
         });
     }
 
+    // Get order row based on _w_id, _d_id, _c_id with the largest _o_id (most recent order)
+    future<> getOrderdByCId() {
+        return getOIdByCIdViaIndex().discard_result()
+                .then([this]() mutable {
+                    return _client.createQuery(tpccCollectionName, "order")
```
[the remark this answers is not in the document]
Comment (ammuv) [reply src/k2/cmd/tpcc/transactions.h:727]: done.

File: src/k2/cmd/tpcc/transactions.h [hunk 3]
```diff
@@ -377,7 +377,13 @@ class NewOrderT : public TPCCTxn
             makeOrderLines();
 
             // Write Order row
-            future<WriteResult> order_update = writeRow<Order>(_order, _txn);
+            future<WriteResult> order_update = writeRow<Order>(_order, _txn).discard_result()
+                                               .then([this]() mutable {
+                                                    // insert secondary index idx_order_customer
+                                                    auto idx_order_customer = IdxOrderCustomer(_order.WarehouseID.value(), _order.DistrictID.value(),
+                                                        _order.CustomerID.value(), _order.OrderID.value());
+                                                    return writeRow<IdxOrderCustomer>(idx_order_customer, _txn);
```
[the remark this answers is not in the document]
Comment (ammuv) [reply src/k2/cmd/tpcc/transactions.h:385]: done.

File: src/k2/cmd/tpcc/transactions.h [hunk 4]
```diff
@@ -678,55 +684,85 @@ class OrderStatusT : public TPCCTxn
...
+            _query_oid.endScanRecord.serializeNext<int16_t>(_d_id);
+            _query_oid.endScanRecord.serializeNext<int32_t>(_c_id);
+            _query_oid.setLimit(1);
+            _query_oid.setReverseDirection(true);
+
+            std::vector<String> projection{"OID"}; // make projection
+            _query_oid.addProjection(projection);
+
+            return _txn.query(_query_oid)
             .then([this] (auto&& response) {
```
[the remark this answers is not in the document]
Comment (ammuv) [reply src/k2/cmd/tpcc/transactions.h:710]: done.

**Review by jfunston — APPROVED** [review]

**Pull request merged**